src/librepgp/stream-dump.c

```diff
@@ -0,0 +1,1235 @@
+            goto finish;
+        }
+
+        tag = get_packet_type(hdr[0]);
+        off = src->readb;
+
+        plen = 0;
+        if (stream_partial_pkt_len(src)) {
+            snprintf(msg, sizeof(msg), "partial len");
+        } else if (stream_intedeterminate_pkt_len(src)) {
+            snprintf(msg, sizeof(msg), "indeterminate len");
+        } else {
+            plen = get_pkt_len(hdr);
+            snprintf(msg, sizeof(msg), "len %zu", plen);
+        }
+        vsnprinthex(smsg, sizeof(smsg), hdr, hlen);
+        dst_printf(dst, ":off %zu: packet header 0x%s (tag %d, %s)\n", off, smsg, tag, msg);
+
+        if (ctx->dump_packets) {
+            ssize_t rlen = plen + hlen;
+            bool    part = false;
+
+            if (!plen || (rlen > 1024 + hlen)) {
+                rlen = 1024 + hlen;
+                part = true;
+            }
+
+            dst_printf(dst, ":off %zu: packet contents ", off + hlen);
+            rlen = src_peek(src, msg, rlen);
+            if (rlen < 0) {
+                dst_printf(dst, "- failed to read\n");
+            } else {
+                rlen -= hlen;
+                if (part || ((size_t) rlen < plen)) {
+                    dst_printf(dst, "(first %d bytes)\n", (int) rlen);
+                } else {
+                    dst_printf(dst, "(%d bytes)\n", (int) rlen);
+                }
+                indent_dest_increase(dst);
+                dst_hexdump(dst, (uint8_t *) msg + hlen, rlen);
+                indent_dest_decrease(dst);
+            }
+            dst_printf(dst, "\n");
+        }
+
+        switch (tag) {
+        case PGP_PTAG_CT_SIGNATURE:
+            ret = stream_dump_signature(ctx, src, dst);
+            break;
+        case PGP_PTAG_CT_SECRET_KEY:
+        case PGP_PTAG_CT_PUBLIC_KEY:
+        case PGP_PTAG_CT_SECRET_SUBKEY:
+        case PGP_PTAG_CT_PUBLIC_SUBKEY:
+            ret = stream_dump_key(ctx, src, dst);
+            break;
+        case PGP_PTAG_CT_USER_ID:
+        case PGP_PTAG_CT_USER_ATTR:
+            ret = stream_dump_userid(src, dst);
+            break;
+        case PGP_PTAG_CT_PK_SESSION_KEY:
+            ret = stream_dump_pk_session_key(ctx, src, dst);
+            break;
+        case PGP_PTAG_CT_SK_SESSION_KEY:
+            ret = stream_dump_sk_session_key(src, dst);
+            break;
+        case PGP_PTAG_CT_SE_DATA:
+        case PGP_PTAG_CT_SE_IP_DATA:
+        case PGP_PTAG_CT_AEAD_ENCRYPTED:
+            ret = stream_dump_encrypted(src, dst, tag);
+            break;
+        case PGP_PTAG_CT_1_PASS_SIG:
+            ret = stream_dump_one_pass(src, dst);
+            break;
+        case PGP_PTAG_CT_COMPRESSED:
+            ret = stream_dump_compressed(ctx, src, dst);
+            break;
+        case PGP_PTAG_CT_LITDATA:
+            ret = stream_dump_literal(src, dst);
+            break;
+        case PGP_PTAG_CT_MARKER:
+        case PGP_PTAG_CT_TRUST:
+        case PGP_PTAG_CT_MDC:
+        default:
```

Comment:
It would be slightly nicer to handle these 3 (marker/trust/mdc) separately from the unknown case. It could print something like "Skipping unhandled pkt: %d".

Reply:
Okay, makes sense.